File: code/SoundPlayground/SoundPlayground/Parser/AbstractSyntaxTree/ContextModifiers/InstrumentBlockModifier.cs
namespace SoundPlayground.Parser.AbstractSyntaxTree.ContextModifiers {
    public class InstrumentBlockModifier : BlockContextModifierNode
    {
        public string InstrumentName { get; set; }

        public InstrumentBlockModifier ( MusicNode body, string instrumentName ) : base( body ) {
            InstrumentName = instrumentName;
        }

        public override void Modify ( Context context )
        {
            if ( !context.Instruments.ContainsKey( InstrumentName ) ) {
                throw new System.Exception( $"Could not switch to non-existent instrument {InstrumentName}" );
            }

            var instrument = context.Instruments[ InstrumentName ];

            if ( instrument.Channel == null ) {
                context.Shared.RegisterInstrument( instrument );
            }

            instrument.RefCount++;

            context.Channel = instrument.Channel.Value;
        }

        public override void Restore ( Context context )
        {
            var instrument = context.Instruments[ InstrumentName ];

            instrument.RefCount--;

            if ( instrument.RefCount == 0 ) {
                context.Shared.UnregisterInstrument( instrument );
            }
        }
    }
}

File: code/SoundPlaygroundPy/parser/abstract_syntax_tree/context_modifiers/instrument_block_modifier.py
from .block_context_modifier_node import BlockContextModifierNode
from core.events import ProgramChangeEvent

class InstrumentBlockModifier( BlockContextModifierNode ):
    def __init__ ( self, body, instrument_name ):
        super().__init__( body )

        self.instrument_name = instrument_name


    def modify ( self, context ):
        instrument = context.symbols.lookup_instrument( self.instrument_name )

        if instrument == None:
            raise BaseException( f"Could not switch to non-existent instrument { self.instrument_name }" )
        
        if instrument.channel == None:
            context.shared.register_instrument( instrument )

            yield ProgramChangeEvent( context.cursor, instrument.channel, instrument.program )

        instrument.ref_count += 1

        context.channel = instrument.channel

    def restore ( self, context ):
        instrument = context.symbols.lookup_instrument( self.instrument_name )

        instrument.ref_count -= 1

        if instrument.ref_count == 0:
            context.shared.unregister_instrument( instrument )
